filter.go

```diff
@@ -0,0 +1,206 @@
+	"net"
+	"strings"
+	"time"
+)
+
+type filter struct {
+	callbacks api.FilterCallbackHandler
+	config    *config
+}
+
+func parseUsernameAndPassword(auth string) (username, password string, ok bool) {
+	const prefix = "Basic "
+	if len(auth) < len(prefix) || !strings.EqualFold(auth[:len(prefix)], prefix) {
+		return "", "", false
+	}
+	c, err := base64.StdEncoding.DecodeString(auth[len(prefix):])
+	if err != nil {
+		return "", "", false
+	}
+	cs := string(c)
+	username, password, ok = strings.Cut(cs, ":")
+	if !ok {
+		return "", "", false
+	}
+	return username, password, true
+}
+
+func dial(config *config) (*ldap.Conn, error) {
+	return ldap.DialURL(
+		// TODO: support TLS
+		fmt.Sprintf("ldap://%s:%d", config.host, config.port),
+		ldap.DialWithDialer(&net.Dialer{
+			Timeout: time.Duration(config.timeout) * time.Second,
+		}),
+	)
+}
+
+// newLdapClient creates a new ldap client.
+func newLdapClient(config *config) (*ldap.Conn, error) {
+	client, err := dial(config)
+	if err != nil {
+		fmt.Println("ldap dial error: ", err)
+		return nil, err
+	}
+
+	err = client.Bind(config.bindDN, config.password)
+	// First bind with a read only user
+	if err != nil {
+		fmt.Println("bind with read only user error: ", err)
+		return nil, err
+	}
+	return client, err
+}
+
+// authLdap authenticates the user against the ldap server.
+func authLdap(config *config, username, password string) bool {
+	if config.filter != "" {
+		return searchMode(config, username, password)
+	}
+
+	// run with bind mode
+	client, err := dial(config)
+	if err != nil {
+		fmt.Println("ldap dial error: ", err)
+		return false
+	}
+
+	_, err = client.SimpleBind(&ldap.SimpleBindRequest{
+		Username: fmt.Sprintf(config.attribute+"=%s,%s", username, config.baseDN),
+		Password: password,
+	})
+	return err == nil
+}
+
+func searchMode(config *config, username, password string) (auth bool) {
+	client, err := newLdapClient(config)
+	if err != nil {
+		fmt.Println("ldap dial error: ", err)
+		return
+	}
+	defer func() {
+		if client != nil {
+			client.Close()
+		}
+		err := recover()
+		if err != nil {
+			auth = false
+			return
+		}
+	}()
+
+	req := ldap.NewSearchRequest(config.baseDN,
+		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+		fmt.Sprintf(config.filter, username),
+		[]string{config.attribute}, nil)
+
+	sr, err := client.Search(req)
+	if err != nil {
+		fmt.Println("ldap search error: ", err)
+		return
+	}
+
+	if len(sr.Entries) != 1 {
+		fmt.Println("ldap search not found: ", err)
+		return
+	}
+
+	userDn := sr.Entries[0].DN
+	err = client.Bind(userDn, password)
+	if err != nil {
+		fmt.Println("ldap bind error: ", err)
+		return
+	}
+
+	auth = true
+	return
+}
+
+func (f *filter) verify(header api.RequestHeaderMap) (bool, string) {
+	auth, ok := header.Get("authorization")
+	if !ok {
+		return false, "no Authorization"
+	}
+	if f.config.cacheTTL > 0 {
+		if _, err := f.config.cache.Get(auth); err == nil {
+			fmt.Printf("cache hit, auth: %v\n", auth)
```

Comment:
this debug code should be removed.
And better add to TODO comment: "add a metrics for it, when the Envoy Golang filter support adding metrics dynamically".